File: frontend/src/lib/student-age.ts
import type { Student } from "@/types";

function calendarParts(value: string): [number, number, number] | null {
  const [year, month, day] = value.split("-").map(Number);
  return Number.isInteger(year) && Number.isInteger(month) && Number.isInteger(day)
    ? [year, month, day]
    : null;
}

export function calendarAge(dateOfBirth: string | null | undefined, referenceDate: string) {
  if (!dateOfBirth) return null;
  const birth = calendarParts(dateOfBirth);
  const reference = calendarParts(referenceDate);
  if (!birth || !reference) return null;

  const [birthYear, birthMonth, birthDay] = birth;
  const [referenceYear, referenceMonth, referenceDay] = reference;
  const birthdayHasPassed =
    referenceMonth > birthMonth || (referenceMonth === birthMonth && referenceDay >= birthDay);
  return referenceYear - birthYear - (birthdayHasPassed ? 0 : 1);
}

export function isMinorOnDate(dateOfBirth: string | null | undefined, referenceDate: string) {
  const age = calendarAge(dateOfBirth, referenceDate);
  return age !== null && age < 18;
}

export function withCurrentMinorStatus(student: Student, referenceDate: string): Student {
  return {
    ...student,
    is_minor: isMinorOnDate(student.date_of_birth, referenceDate),
  };
}

File: backend/app/services/student_age.py
from datetime import date
from typing import Optional


def age_on_date(date_of_birth: date, reference_date: date) -> int:
    age = reference_date.year - date_of_birth.year
    if (reference_date.month, reference_date.day) < (
        date_of_birth.month,
        date_of_birth.day,
    ):
        age -= 1
    return age


def is_minor_on_date(date_of_birth: Optional[date | str], reference_date: date) -> bool:
    if not date_of_birth:
        return False
    if isinstance(date_of_birth, str):
        date_of_birth = date.fromisoformat(date_of_birth)
    return age_on_date(date_of_birth, reference_date) < 18
